### PR TITLE
[TensorExpr] Simplify TE IR before applying any transformations.

### PR DESCRIPTION
**Pull request by ZolotukhinM**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #64717

This also exposed several bugs, which are fixed in this PR.

Differential Revision: [D30826408](https://our.internmc.facebook.com/intern/diff/D30826408)

### CODE REVIEW

**Comment by ZolotukhinM** [comment]
@ZolotukhinM has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30826408).

**Comment by ZolotukhinM** [comment]
@ZolotukhinM has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30826408).

File: test/cpp/tensorexpr/test_kernel.cpp [hunk 1]
```diff
@@ -210,7 +210,11 @@ TEST_F(Kernel, Huge) {
   TensorExprKernel k(graph);
   std::ostringstream oss;
   oss << *k.getCodeGenStmt();
-  const std::string& verification_pattern = "# CHECK: 4000000000";
+  // The 4000000000 iterations loop will be split into 500000000 x 8 and the
+  // outer loop will be parallel. If LLVM is not present, it will not be split,
+  // and to cover both of these cases we're looking for 00000000ll; in the
+  // output.
+  const std::string& verification_pattern = R"IR(# CHECK: 00000000ll;)IR";
```
Comment (navahgar) [inline test/cpp/tensorexpr/test_kernel.cpp:217]: nit: we don't quite have to check for that number. We just want to make sure that the compilation is successful here. So, even checking for something like `aten_relu` would suffice..

File: torch/csrc/jit/tensorexpr/loopnest.cpp [hunk 2]
```diff
@@ -731,6 +731,9 @@ bool LoopNest::computeInline(StmtPtr s) {
 bool LoopNest::computeInline(BufPtr b) {
   // If buf is used or defined in an ExternalCall, we cannot inline it
   auto buf_load_store_uses = findLoadOrStoreUses(root_stmt_);
+  if (!buf_load_store_uses.count(b)) {
```
Comment (navahgar) [inline torch/csrc/jit/tensorexpr/loopnest.cpp:734]: When does this case happen? Just curious. Also, is this relevant / required for this PR?

**Review by navahgar — APPROVED** [review]

File: test/cpp/tensorexpr/test_kernel.cpp (hunk 1 above)
Comment (ZolotukhinM) [reply test/cpp/tensorexpr/test_kernel.cpp:217]: I wanted to still have a check for a number to make sure we're not truncating it or something. But if it happens to be too fragile, we can indeed just check that it compiles.

File: torch/csrc/jit/tensorexpr/loopnest.cpp (hunk 2 above)
Comment (ZolotukhinM) [reply torch/csrc/jit/tensorexpr/loopnest.cpp:734]: Yes, this is required, and it happens in one of our tests where we concat empty tensors. Before I added simplifier, we had a stmt like
```
{
for (i = 0; i < 0; i++)
  a[i] = 0
}
```
With simplifier the entire block becomes empty.

**Comment by ZolotukhinM** [comment]
@ZolotukhinM has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30826408).

**Comment by ZolotukhinM** [comment]
@ZolotukhinM has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30826408).

**Comment by ZolotukhinM** [comment]
@ZolotukhinM has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30826408).

**Pull request closed without merging**